drape_frontend/arrow3d.cpp

```diff
@@ -5,71 +5,365 @@
+    callbacks.file_read = fast_obj_adapter::FileRead;
+    callbacks.file_size = fast_obj_adapter::FileSize;
+    meshData = fast_obj_read_with_callbacks(kMainFileId.c_str(), &callbacks, &source);
+    CHECK(meshData != nullptr, ());
+    
+    // Fill buffers.
+    std::vector<float> positions;
+    if (meshData->position_count > 1)
+      positions.resize(meshData->index_count * kComponentsInVertex);
+    
+    std::vector<float> normals;
+    if (meshData->normal_count > 1)
+      normals.resize(meshData->index_count * kComponentsInNormal);
+    
+    std::vector<float> texCoords;
+    if (meshData->texcoord_count > 1)
+      texCoords.resize(meshData->index_count * kComponentsInTexCoord);
+    
+    for (uint32_t i = 0; i < meshData->index_count; ++i)
+    {
+      if (meshData->position_count > 1)
+      {
+        memcpy(&positions[i * kComponentsInVertex],
+               &meshData->positions[meshData->indices[i].p * kComponentsInVertex],
+               sizeof(float) * kComponentsInVertex);
+      }
+      
+      if (meshData->normal_count > 1)
+      {
+        memcpy(&normals[i * kComponentsInNormal],
+               &meshData->normals[meshData->indices[i].n * kComponentsInNormal],
+               sizeof(float) * kComponentsInNormal);
+      }
+      
+      if (meshData->texcoord_count > 1)
+      {
+        memcpy(&texCoords[i * kComponentsInTexCoord],
+               &meshData->texcoords[meshData->indices[i].t * kComponentsInTexCoord],
+               sizeof(float) * kComponentsInTexCoord);
+      }
+    }
+    
+    completionHandler(std::move(positions), std::move(normals), std::move(texCoords));
+    
+    fast_obj_destroy(meshData);
+  }
+  catch (RootException & e)
+  {
+    if (meshData)
+      fast_obj_destroy(meshData);
+    failureHandler(e.what());
+    return false;
+  }
 
-  std::vector<float> normals =
-      dp::MeshObject::GenerateNormalsForTriangles(vertices, kComponentsInVertex);
+  return true;
+}
+}  // namespace
 
-  auto constexpr kVerticesBufferInd = 0;
-  auto copiedVertices = vertices;
-  m_arrowMesh.SetBuffer(kVerticesBufferInd, std::move(copiedVertices),
-                        sizeof(float) * kComponentsInVertex);
-  m_arrowMesh.SetAttribute("a_pos", kVerticesBufferInd, 0 /* offset */, kComponentsInVertex);
+// static
+Arrow3d::PreloadedData Arrow3d::PreloadMesh(std::optional<Arrow3dCustomDecl> const & customDecl,
+                                            ref_ptr<dp::TextureManager> texMng)
+{
+  Arrow3d::PreloadedData data;
+  
+  bool const useDefaultResource = !customDecl.has_value() ||
+                                  customDecl->m_loadFromDefaultResourceFolder;
+  
+  // Load arrow mesh.
+  std::string meshPath = customDecl.has_value() ? customDecl->m_arrowMeshPath :
+                                                  kDefaultArrowMesh;
+  data.m_meshData = PreloadedMeshData{};
+  if (!LoadMesh(meshPath, useDefaultResource, [&](std::vector<float> positions,
```

Comment:
Would && help here to avoid unnecessary vector copies?

Reply:
Not really, vectors are moved
```
completionHandler(std::move(positions), std::move(normals), std::move(texCoords));
```